core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala

```diff
@@ -239,16 +266,26 @@ protected[core] abstract class EntitlementProvider(config: WhiskConfig, loadBala
      * If the requested right is an activation, the set of resources must contain an activation of an action or filter to be throttled.
      * While it is possible for the set of resources to contain more than one action or trigger, the plurality is ignored and treated
      * as one activation since these should originate from a single macro resources (e.g., a sequence).
+     *
+     * @param subject the subject to check quota for
+     * @param right the privilege, if ACTIVATE then check quota else return None
+     * @param resource the set of resources must contain at least one resource that can be activated else return None
+     * @return None if subject is not throttled else a rejection
      */
-    protected def checkUserThrottle(subject: Subject, right: Privilege, resources: Set[Resource])(implicit transid: TransactionId) = {
+    private def checkUserThrottle(subject: Subject, right: Privilege, resources: Set[Resource])(
+        implicit transid: TransactionId): Option[RejectRequest] = {
         def userThrottled = {
             val isInvocation = resources.exists(_.collection.path == Collection.ACTIONS)
             val isTrigger = resources.exists(_.collection.path == Collection.TRIGGERS)
             (isInvocation && !invokeRateThrottler.check(subject)) || (isTrigger && !triggerRateThrottler.check(subject))
         }
+        hasTooManyRequests(right == ACTIVATE && userThrottled)
+    }
 
-        if (right == ACTIVATE && userThrottled) {
-            Some(Future.failed(RejectRequest(TooManyRequests, tooManyRequests)))
+    /** Helper. */
+    private def hasTooManyRequests(yes: Boolean)(implicit transid: TransactionId) = {
+        if (yes) {
+            Some(RejectRequest(TooManyRequests, tooManyRequests))
```

Comment:
As noted in person:
This feels rather odd, because it makes `Option.map` follow the "unhappy" path, while it usually reads a being the positive path. Further above you `map` into `Future.failed` which seems strange to read.